src/compressor/QatAccel.cc

```diff
@@ -11,67 +11,100 @@
  * Foundation.  See file COPYING.
  *
  */
-
+#include <qatzip.h>
 #include "QatAccel.h"
 
+#define SESSIONS_NUM 200
+
+void QzSessionDeleter::operator() (struct QzSession_S *session) {
+  if (NULL != session->internal) {
+    qzTeardownSession(session);
+    qzClose(session);
+  }
+  delete session;
+}
+
 /* Estimate data expansion after decompression */
 static const unsigned int expansion_ratio[] = {5, 20, 50, 100, 200};
 
-QatAccel::~QatAccel() {
-  if (NULL != session.internal) {
-    qzTeardownSession(&session);
-    qzClose(&session);
+QatAccel::QatAccel() {
+  sessions.resize(SESSIONS_NUM);
+  for (auto &session : sessions) {
+    session.reset(new struct QzSession_S);
+    memset(session.get(), 0, sizeof(struct QzSession_S));
   }
 }
 
+QatAccel::~QatAccel() {}
+
 bool QatAccel::init(const std::string &alg) {
   QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
   int rc;
 
+  std::unique_lock<std::mutex> l(lock);
+  if (!alg.empty()) {
+    return true;
+  }
+
   rc = qzGetDefaults(&params);
   if (rc != QZ_OK)
     return false;
   params.direction = QZ_DIR_BOTH;
-  if (alg == "snappy")
-    params.comp_algorithm = QZ_SNAPPY;
-  else if (alg == "zlib")
+  if (alg == "zlib")
     params.comp_algorithm = QZ_DEFLATE;
-  else if (alg == "lz4")
-    params.comp_algorithm = QZ_LZ4;
   else
     return false;
 
   rc = qzSetDefaults(&params);
   if (rc != QZ_OK)
       return false;
 
-  rc = qzInit(&session, QZ_SW_BACKUP_DEFAULT);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
-    return false;
+  for (auto &session : sessions) {
+    rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
+    if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+      return false;
 
-  rc = qzSetupSession(&session, &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(&session);
-    qzClose(&session);
-    return false;
+    rc = qzSetupSession(session.get(), &params);
+    if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
+      qzTeardownSession(session.get());
+      qzClose(session.get());
+      return false;
+    }
   }
 
+  this->alg = alg;
   return true;
 }
 
 int QatAccel::compress(const bufferlist &in, bufferlist &out, boost::optional<int32_t> &compressor_message) {
+  unique_session_ptr session;
+  {
+    std::unique_lock<std::mutex> l(lock);
+    if (sessions.empty()) {
+      conds.wait(l, [this](){return !sessions.empty();});
+    }
```

Comment:
if `sessions` are set in the ctor of `QatAccel`, why do we need to check for empty in a member function?

 ```suggestion
    conds.wait(l, [this] { return !sessions.empty(); });
```

would suffice. please check the document of `condition_variable` and implementation of this class for more details.

Reply:
yes, `conds.wait()` has a judgment for the first time. Accept it.